x-pack/apm-server/sampling/eventstorage/storage.go

```diff
@@ -0,0 +1,228 @@
+func (s *Storage) NewShardedReadWriter() *ShardedReadWriter {
+	return newShardedReadWriter(s)
+}
+
+// NewReadWriter returns a new ReadWriter for reading events from and
+// writing events to storage.
+//
+// The returned ReadWriter must be closed when it is no longer needed.
+func (s *Storage) NewReadWriter() *ReadWriter {
+	return &ReadWriter{
+		s:   s,
+		txn: s.db.NewTransaction(true),
+	}
+}
+
+// ReadWriter provides a means of reading events from storage, and batched
+// writing of events to storage.
+//
+// ReadWriter is not safe for concurrent access. All operations that involve
+// a given trace ID should be performed with the same ReadWriter in order to
+// avoid conflicts, e.g. by using consistent hashing to distribute to one of
+// a set of ReadWriters, such as implemented by ShardedReadWriter.
+type ReadWriter struct {
+	s             *Storage
+	txn           *badger.Txn
+	pendingWrites int
+
+	// readKeyBuf is a reusable buffer for keys used in read operations.
+	// This must not be used in write operations, as keys are expected to
+	// be unmodified until the end of a transaction.
+	readKeyBuf []byte
+}
+
+// Close closes the writer. Any writes that have not been flushed may be lost.
+//
+// This must be called when the writer is no longer needed, in order to reclaim
+// resources.
+func (rw *ReadWriter) Close() {
+	rw.txn.Discard()
+}
+
+// Flush waits for preceding writes to be committed to storage.
+//
+// Flush must be called to ensure writes are committed to storage.
+// If Flush is not called before the writer is closed, then writes
+// may be lost.
+func (rw *ReadWriter) Flush() error {
+	err := rw.txn.Commit()
+	rw.txn = rw.s.db.NewTransaction(true)
+	rw.pendingWrites = 0
+	return err
+}
+
+// WriteTraceSampled records the tail-sampling decision for the given trace ID.
+func (rw *ReadWriter) WriteTraceSampled(traceID string, sampled bool) error {
+	key := []byte(traceID)
+	var meta uint8 = entryMetaTraceUnsampled
+	if sampled {
+		meta = entryMetaTraceSampled
+	}
+	entry := badger.NewEntry(key[:], nil).WithMeta(meta)
+	return rw.writeEntry(entry.WithTTL(rw.s.ttl))
+}
+
+// IsTraceSampled reports whether traceID belongs to a trace that is sampled
+// or unsampled. If no sampling decision has been recorded, IsTraceSampled
+// returns ErrNotFound.
+func (rw *ReadWriter) IsTraceSampled(traceID string) (bool, error) {
+	rw.readKeyBuf = append(rw.readKeyBuf[:0], traceID...)
+	item, err := rw.txn.Get(rw.readKeyBuf)
+	if err != nil {
+		if err == badger.ErrKeyNotFound {
+			return false, ErrNotFound
+		}
+		return false, err
+	}
+	return item.UserMeta() == entryMetaTraceSampled, nil
+}
+
+// WriteTransaction writes tx to storage.
+//
+// WriteTransaction may return before the write is committed to storage.
+// Call Flush to ensure the write is committed.
+func (rw *ReadWriter) WriteTransaction(tx *model.Transaction) error {
+	if tx.Sampled != nil && !*tx.Sampled {
+		return errors.New("transaction is not sampled")
+	}
```

Comment:
Why is this not written if transactio is not sampled; should this be the concern of this method?

Reply:
We should only be writing events that were sampled by the agent. But you're right, it's not really the concern of this method, I'll remove it.